packages/assets-controllers/src/TokenDetectionController.ts

```diff
@@ -381,13 +401,33 @@ export class TokenDetectionController extends StaticIntervalPollingController<
     }, this.getIntervalLength());
   }
 
-  #getCorrectChainId(networkClientId?: NetworkClientId) {
-    const { chainId } =
-      this.messagingSystem.call(
+  #getCorrectChainIdAndNetworkClientId(networkClientId?: NetworkClientId): {
+    chainId: Hex;
+    networkClientId: NetworkClientId;
+  } {
+    if (networkClientId) {
+      const networkConfiguration = this.messagingSystem.call(
         'NetworkController:getNetworkConfigurationByNetworkClientId',
-        networkClientId ?? this.#networkClientId,
-      ) ?? {};
-    return chainId ?? this.#chainId;
+        networkClientId,
+      );
+      if (networkConfiguration) {
+        return {
+          chainId: networkConfiguration.chainId,
+          networkClientId,
+        };
+      }
+    }
+    const { chainId } = this.messagingSystem.call(
+      'NetworkController:getProviderConfig',
+    );
+    const newNetworkClientId = this.messagingSystem.call(
+      'NetworkController:findNetworkClientIdByChainId',
+      this.#chainId,
+    );
```

Comment:
What do you think about replacing the `this.#chainId` with the `chainId` from `getProviderConfig`? https://github.com/MetaMask/core/pull/3923/commits/eccf79c7014f23e8bf94e037d1a5b657539a761f

Since the purpose of this method is to correct cases where the cached `this.#chainId` or `this.#networkClientId` are stale, having it rely solely on the network-controller as its source of truth feels better to me.

Reply:
I agree that would help, although we could go even further and not use `providerConfig` at all, so that we wouldn't have to use `findNetworkClientIdByChainId` (which, as we've seen, can give us something we don't expect). Instead we could get `selectedNetworkClientId` from NetworkController state and then call `getNetworkConfigurationByNetworkClientId` to get the chain ID associated with that network client. Maybe we save that for a new PR

Reply:
I went ahead and implemented your suggestion in the refactor PR: https://github.com/MetaMask/core/pull/3938/commits/c75fb3bd37ddf2715ed250bdd2c2331a54008b72. 

I used `getNetworkClientById` instead of `getNetworkConfigurationByNetworkClientId` to avoid having to use `this.#chainId` as a fallback value for the case where a `undefined` network configuration is returned.